File: projects/interview_query/first_unique_character.clj
(def word-1 "concussion")
(def word-2 "helloworld")

(defn ->string [characters]
  (map str characters))

(defn create-schema [characters]
  (letfn [(->schema [index character]
            {:index index :character character})]
    (map-indexed ->schema characters)))

(defn count-characters [schemas]
  (group-by :character schemas))

(defn filter-by-lower-frequency [grouped-schemas]
  (->> grouped-schemas
       vals
       (filter #(= 1 (count %)))))

(defn sort-by-index [schemas]
  (sort-by :index schemas))

(defn get-index [first-schema]
  (get-in first-schema [0 :index]))

(defn first-unique-character [word]
  (-> word
      vec
      ->string
      create-schema
      count-characters
      filter-by-lower-frequency
      sort-by-index
      first
      get-index))

(-> word-1 first-unique-character print)

File: algorithms/interview_query/first_unique_character.rb
# frozen_string_literal: true

word = 'concussion'

def create_schemas(characters)
  characters.map.with_index { |character, index| { index: index, character: character } }
end

def count_characters(schemas)
  schemas.group_by { |schema| schema[:character] }
end

def filter_by_lower_frequency(grouped_schemas)
  grouped_schemas.values.filter { |group| group.length == 1 }
end

def sort_by_index(filtered_schemas)
  filtered_schemas.flatten.sort_by { |schema| schema[:index] }
end

def get_index(sorted_schemas)
  sorted_schemas.first[:index]
end

def first_unique_character(word)
  word
    .chars
    .then { |characters| create_schemas(characters) }
    .then { |schemas| count_characters(schemas) }
    .then { |grouped_schemas| filter_by_lower_frequency(grouped_schemas) }
    .then { |filtered_schemas| sort_by_index(filtered_schemas) }
    .then { |sorted_schemas| get_index(sorted_schemas) }
end

puts first_unique_character(word)
